rust/lance-index/src/scalar/bitmap.rs

```diff
@@ -323,45 +339,95 @@ fn get_batch_from_arrays(
+        {
+            let keys_array = if cur_keys.is_empty() {
+                new_empty_array(value_type)
+            } else {
+                ScalarValue::iter_to_array(cur_keys.clone().into_iter()).unwrap()
+            };
+            let mut binary_builder = BinaryBuilder::new();
+            for b in &cur_bitmaps {
+                binary_builder.append_value(b);
+            }
+            let bitmaps_array = Arc::new(binary_builder.finish()) as Arc<dyn Array>;
+            batches.push((keys_array, bitmaps_array));
+            cur_keys.clear();
+            cur_bitmaps.clear();
+            cur_bytes = 0;
+        }
 
-    Arc::new(builder.finish())
+        cur_keys.push(key);
+        cur_bitmaps.push(bytes);
+        cur_bytes += bitmap_len;
+    }
+
+    // Flush any remaining
+    if !cur_keys.is_empty() {
+        let keys_array = ScalarValue::iter_to_array(cur_keys).unwrap();
+        let mut binary_builder = BinaryBuilder::new();
+        for b in &cur_bitmaps {
+            binary_builder.append_value(b);
+        }
+        let bitmaps_array = Arc::new(binary_builder.finish()) as Arc<dyn Array>;
+        batches.push((keys_array, bitmaps_array));
+    }
+
+    batches
 }
 
 async fn write_bitmap_index(
     state: HashMap<ScalarValue, RowIdTreeMap>,
     index_store: &dyn IndexStore,
     value_type: &DataType,
 ) -> Result<()> {
-    let keys_iter = state.keys().cloned();
-    let keys_array = if state.is_empty() {
-        new_empty_array(value_type)
+    let batches = chunked_keys_and_bitmaps(state, value_type);
+
+    // Create a schema even if we have no batches
+    let schema = if batches.is_empty() {
+        // For empty state, create a schema with the same structure
+        Arc::new(Schema::new(vec![
+            Field::new("keys", value_type.clone(), true),
+            Field::new("bitmaps", DataType::Binary, true),
+        ]))
     } else {
-        ScalarValue::iter_to_array(keys_iter)?
+        // Get schema from first batch
+        let (first_keys, first_bitmaps) = &batches[0];
+        Arc::new(Schema::new(vec![
+            Field::new("keys", first_keys.data_type().clone(), true),
+            Field::new("bitmaps", first_bitmaps.data_type().clone(), true),
+        ]))
     };
 
-    let values_iter = state.into_values();
-    let binary_bitmap_array = get_bitmaps_from_iter(values_iter);
-
-    let record_batch = get_batch_from_arrays(keys_array, binary_bitmap_array)?;
-
     let mut bitmap_index_file = index_store
-        .new_index_file(BITMAP_LOOKUP_NAME, record_batch.schema())
+        .new_index_file(BITMAP_LOOKUP_NAME, schema)
         .await?;
-    bitmap_index_file.write_record_batch(record_batch).await?;
+
+    // Write all batches to the same file
+    for (keys_array, binary_bitmap_array) in batches.into_iter() {
+        let record_batch = get_batch_from_arrays(keys_array, binary_bitmap_array)?;
+        bitmap_index_file.write_record_batch(record_batch).await?;
+    }
```

Comment:
As mentioned earlier, it would be ideal to write batches as they are available, rather than buffering them all up and writing them.

Reply:
refactored